Make classes customizable

FCUnlimited opened the pull request:
Declared some functions as protected and added also some optional parameter.

Also made the WPF Image clickable. So if you click an image the associated command will be executed. In my case to zoom into it.

Commit:
- Made the WpfRenderer extendable

Commit:
Make it possible to customize the WPF rendering

Commit:
- Made WPF images clickable

Kryptos-FR commented on src/Markdig.Wpf/Markdown.cs:
```diff
@@ -24,14 +24,48 @@ public static partial class Markdown
         /// <exception cref="System.ArgumentNullException">if markdown variable is null</exception>
         [NotNull]
-        public static FlowDocument ToFlowDocument([NotNull] string markdown, MarkdownPipeline pipeline = null)
+        public static FlowDocument PathToFlowDocument([NotNull] string markdown, MarkdownPipeline pipeline = null, WpfRenderer renderer = null)
```
Not sure to get why you have a differently named method doing the same?

FCUnlimited replied:
Ups, that was a try.
My problem was that usually, Markdownrenderer like Visual Studio Code resolve links from the current document. Your library uses the application base path.
I resolved this in a derived class of the WPFRenderer

Kryptos-FR commented on src/Markdig.Wpf/Renderers/Wpf/Inlines/LinkInlineRenderer.cs:
```diff
@@ -31,14 +31,20 @@ protected override void Write([NotNull] WpfRenderer renderer, [NotNull] LinkInli
-                var image = new Image
+                var template = new ControlTemplate();
+                var image = new FrameworkElementFactory(typeof(Image));
+                image.SetValue(Image.SourceProperty, new System.Windows.Media.Imaging.BitmapImage(new Uri(url, UriKind.RelativeOrAbsolute)));
```
Namespace `System.Windows.Media.Imaging` is already imported, type usage can be simplified.

Kryptos-FR commented on src/Markdig.Wpf/Renderers/Wpf/Inlines/LinkInlineRenderer.cs:
```diff
@@ -31,14 +31,20 @@ protected override void Write([NotNull] WpfRenderer renderer, [NotNull] LinkInli
             {
-                var image = new Image
+                var template = new ControlTemplate();
+                var image = new FrameworkElementFactory(typeof(Image));
```
I have never used that type before (`FrameworkElementFactory`). Out of cursiority why can't you instantiate an image directly (and use it as the content of the `Button`)?

According to the documentation, that class is deprecated (https://docs.microsoft.com/en-us/dotnet/api/system.windows.frameworkelementfactory). But maybe there is no other way?

FCUnlimited replied:
As in Xaml, u have to create a Template. With this FrameworkElementFactory, u create such one.
Another way would be to write Xaml and load this string as shown in the following links:
[Codeproject which says u can do it in most cases](https://www.codeproject.com/Tips/808808/Create-Data-and-Control-Templates-using-Delegates)
[He explains why u shouldn't do it that way](https://www.ikriv.com/dev/wpf/DataTemplateCreation/)
I think its the easiest way to do it with this FrameworkElementFactory as long as it is not possible to provide Templates for all Wpf rendered controls...

Kryptos-FR commented on src/Markdig.Wpf/Renderers/WpfRenderer.cs:
```diff
@@ -31,37 +31,26 @@ public class WpfRenderer : RendererBase
+            LoadDocument(document);
+        }
+
+        public void LoadDocument([NotNull] FlowDocument document)
```
Let's make that one `virtual` as well.

Kryptos-FR commented on src/Markdig.Wpf/Renderers/WpfRenderer.cs:
```diff
@@ -31,37 +31,26 @@ public class WpfRenderer : RendererBase
         }
 
-        public FlowDocument Document { get; }
+        public FlowDocument Document { get; private set; }
```
`protected` if we make `LoadDocument(FlowDocument)` `virtual`

Kryptos-FR reviewed (changes requested):
Looks ok for now (with a few remarks and suggestion). Do you have a custom implementation of the renderer I could try out?

Commit:
- Removed PathToFlowDocument function

Commit:
- LoadDocument now virtual

Commit:
- Always copy BenchmarkCPU.png
- Added ImageComannd example

Commit:
- Added example for customizing classes (#33)

FCUnlimited commented:
I implemented your suggestions and added the "Markdig.Wpf.SampleAppCustomized" project as u wished.
I also changed the copy property of the benchmark image. Now it gets always copied.

Kryptos-FR commented:
Thanks. I'll try to review this week but I'm quite busy and I'm not available this weekend. So it might take a while.

FCUnlimited commented:
> Thanks. I'll try to review this week but I'm quite busy and I'm not available this weekend. So it might take a while.

Would be awesome, if u can publish it on NuGet till the end of the week. If this is possible :)
